### PR TITLE
Remove explicit use of futures in tridiagonal solver

### LINKED ISSUE #588: Remove use of `shared_future` from `tridiag_solver`
Sub-task of #468. See https://github.com/eth-cscs/DLA-Future/pull/436#discussion_r913496374.

### PR DESCRIPTION
**Pull request by msimberg**
Fixes #588. Removes all explicit occurrences of futures, but does not yet get rid of the ones that come implicitly from `Matrix::read` (which returns a keep_future(shared_future)`).

~There's one TODO related to lifetime management of tiles that I haven't yet decided if I'll leave as it is in this PR or try to do something about it. Hence the draft.~ I've left a TODO for something that will be changed separately.

### CODE REVIEW

File: include/dlaf/eigensolver/tridiag_solver/merge.h
Comment (msimberg) [inline include/dlaf/eigensolver/tridiag_solver/merge.h:222]: I've updated the comment here with a link to https://github.com/pika-org/pika/issues/479. A corresponding issue on the DLA-Future side is https://github.com/eth-cscs/DLA-Future/issues/665.

Until those issues are resolved `ensure_started` acts as a hacky replacement for the same purpose and I won't change it in this PR.

**Comment by msimberg** [comment]
This should be pretty much ready to go now. However, @teonnik, if this conflicts with your changes to have the distributed version working I absolutely don't mind having your changes in first and adapting this later again to your distributed changes. Just comment with whatever is easiest for you.

File: include/dlaf/eigensolver/tridiag_solver/merge.h
Comment (teonnik) [reply include/dlaf/eigensolver/tridiag_solver/merge.h:222]: I don't completely understand the issue here, there are no tiles involved in the call to `di::transform()`. Can you explain a bit more?

**Review by teonnik — APPROVED** [review]
LGTM!

> However, @teonnik, if this conflicts with your changes to have the distributed version working I absolutely don't mind having your changes in first and adapting this later again to your distributed changes.

I think we can have it merged. It will not be an issue for the distributed tridiagonal solver.

File: include/dlaf/eigensolver/tridiag_solver/merge.h
Comment (msimberg) [reply include/dlaf/eigensolver/tridiag_solver/merge.h:222]: It maybe doesn't look like it, but the tiles come from `maxVectorElement`. The vectors of `shared_future<Tile>` will be kept for too long in the operation state unless some adaptor explicitly releases the operation states up to this point. `ensure_started` is overkill, but gets the job done (because of how we have implemented it in pika) and is neutral enough to be used like this.

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
This has also failed due to timeouts. I've restarted the jobs.

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
I'm doing one last attempt at CI (the CUDA codecov tests keep not getting an allocation), but I think the failures are all unrelated, so will merge unless something really nasty shows up.

bors try

**Comment by msimberg** [comment]
I've rebased this on #686. I will still check how much this _increases_ the binary sizes...

bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try-

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
This is ready for reviews but as the "on hold" label says this should not be merged yet.

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
One of the test jobs didn't get to start: https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/4700071344751697/7514005670787789/-/jobs/3473824455.

**Comment by msimberg** [comment]
cscs-ci run

**Review by rasolca — APPROVED** [review]

**Pull request merged**
